src/graph/immutable_graph.cc

```diff
@@ -156,392 +185,55 @@ DegreeArray ImmutableGraph::CSR::GetDegrees(IdArray vids) const {
-
-  auto succ = this->GetOutCSR()->GetIndexRef(vid);
-  const int64_t len = succ.size();
-  IdArray rst = IdArray::Empty({len}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
-  dgl_id_t* rst_data = static_cast<dgl_id_t*>(rst->data);
-
-  std::copy(succ.begin(), succ.end(), rst_data);
-  return rst;
-}
-
-DGLIdIters ImmutableGraph::GetInEdgeIdRef(dgl_id_t src, dgl_id_t dst) const {
-  CHECK(this->in_csr_);
-  auto pred = this->in_csr_->GetIndexRef(dst);
-  auto it = std::lower_bound(pred.begin(), pred.end(), src);
-  // If there doesn't exist edges between the two nodes.
-  if (it == pred.end() || *it != src) {
-    return DGLIdIters(it, it);
-  }
-
-  size_t off = it - in_csr_->indices.begin();
-  CHECK(off < in_csr_->indices.size());
-  auto start = in_csr_->edge_ids.begin() + off;
-  int64_t len = 0;
-  // There are edges between the source and the destination.
-  for (auto it1 = it; it1 != pred.end() && *it1 == src; it1++, len++) {}
-  return DGLIdIters(start, start + len);
-}
-
-DGLIdIters ImmutableGraph::GetOutEdgeIdRef(dgl_id_t src, dgl_id_t dst) const {
-  CHECK(this->out_csr_);
-  auto succ = this->out_csr_->GetIndexRef(src);
-  auto it = std::lower_bound(succ.begin(), succ.end(), dst);
-  // If there doesn't exist edges between the two nodes.
-  if (it == succ.end() || *it != dst) {
-    return DGLIdIters(it, it);
-  }
-
-  size_t off = it - out_csr_->indices.begin();
-  CHECK(off < out_csr_->indices.size());
-  auto start = out_csr_->edge_ids.begin() + off;
-  int64_t len = 0;
-  // There are edges between the source and the destination.
-  for (auto it1 = it; it1 != succ.end() && *it1 == dst; it1++, len++) {}
-  return DGLIdIters(start, start + len);
-}
-
-IdArray ImmutableGraph::EdgeId(dgl_id_t src, dgl_id_t dst) const {
-  CHECK(HasVertex(src) && HasVertex(dst)) << "invalid edge: " << src << " -> " << dst;
-
-  auto edge_ids = in_csr_ ? GetInEdgeIdRef(src, dst) : GetOutEdgeIdRef(src, dst);
-  int64_t len = edge_ids.size();
-  IdArray rst = IdArray::Empty({len}, DLDataType{kDLInt, 64, 1}, DLContext{kDLCPU, 0});
-  dgl_id_t* rst_data = static_cast<dgl_id_t*>(rst->data);
-  if (len > 0) {
-    std::copy(edge_ids.begin(), edge_ids.end(), rst_data);
+IdArray CSR::EdgeId(dgl_id_t src, dgl_id_t dst) const {
+  CHECK(HasVertex(src)) << "invalid vertex: " << src;
+  CHECK(HasVertex(dst)) << "invalid vertex: " << dst;
+  std::vector<dgl_id_t> ids;
+  const dgl_id_t* indptr_data = static_cast<dgl_id_t*>(indptr_->data);
+  const dgl_id_t* indices_data = static_cast<dgl_id_t*>(indices_->data);
+  const dgl_id_t* eid_data = static_cast<dgl_id_t*>(edge_ids_->data);
+  for (dgl_id_t i = indptr_data[src]; i < indptr_data[src+1]; ++i) {
+    if (indices_data[i] == dst) {
+      ids.push_back(eid_data[i]);
+    }
   }
-
-  return rst;
+  return VecToIdArray(ids);
 }
 
-ImmutableGraph::EdgeArray ImmutableGraph::EdgeIds(IdArray src_ids, IdArray dst_ids) const {
+CSR::EdgeArray CSR::EdgeIds(IdArray src_ids, IdArray dst_ids) const {
```

Comment:
it seems this operation is heavily used by the computation API, for example `send_and_recv`. I think it's important to optimize its performance. because the edges in the neighborhood of a vertex aren't sorted, we need linearly search in the neighborhood. it can potentially be much slower than the previous implementation.

Reply:
Yeah. Do you have any benchmark I could refer to? As we've discussed before, I could implement the `sorted_indices` if that's really needed.

Reply:
I don't unfortunately. But we could create a relatively large graph and run `send_and_recv` to benchmark it.

Reply:
I put down a TODO note here. Let's do that later?

Reply:
sure